nixos/modules/tasks/filesystems/overlayfs.nix

```diff
@@ -21,89 +33,111 @@ let
+          RequiresMountsFor = "${upperdir} ${workdir}";
+        };
+        serviceConfig = {
+          Type = "oneshot";
+          ExecStart = "${pkgs.coreutils}/bin/mkdir -p -m 0755 ${upperdir} ${workdir}";
         };
       };
+    };
 
-  overlayOpts = { config, ... }: {
+  overlayOpts =
+    { config, ... }:
+    {
 
-    options.overlay = {
+      options.overlay = {
 
-      lowerdir = lib.mkOption {
-        type = with lib.types; nullOr (nonEmptyListOf (either str pathInStore));
-        default = null;
-        description = ''
-          The list of path(s) to the lowerdir(s).
+        lowerdir = lib.mkOption {
+          type = with lib.types; nullOr (nonEmptyListOf (either str pathInStore));
+          default = null;
+          description = ''
+            The list of path(s) to the lowerdir(s).
 
-          To create a writable overlay, you MUST provide an upperdir and a
-          workdir.
+            To create a writable overlay, you MUST provide an `upperdir` and a
+            `workdir`.
 
-          You can create a read-only overlay when you provide multiple (at
-          least 2!) lowerdirs and neither an upperdir nor a workdir.
-        '';
-      };
+            You can create a read-only overlay when you provide multiple (at
+            least 2!) lowerdirs and neither an `upperdir` nor a `workdir`.
+          '';
+        };
 
-      upperdir = lib.mkOption {
-        type = lib.types.nullOr lib.types.str;
-        default = null;
-        description = ''
-          The path to the upperdir.
+        upperdir = lib.mkOption {
+          type = lib.types.nullOr lib.types.str;
+          default = null;
+          description = ''
+            The path to the upperdir.
 
-          If this is null, a read-only overlay is created using the lowerdir.
+            If this is null, a read-only overlay is created using the lowerdir.
 
-          If you set this to some value you MUST also set `workdir`.
-        '';
-      };
+            If the filesystem is `neededForBoot`, this will be prefixed with `/sysroot`,
+            unless `underlyingDirectoriesOnRealRoot` is set to `true`.
+
+            If you set this to some value you MUST also set `workdir`.
+          '';
+        };
+
+        workdir = lib.mkOption {
+          type = lib.types.nullOr lib.types.str;
+          default = null;
+          description = ''
+            The path to the workdir.
 
-      workdir = lib.mkOption {
-        type = lib.types.nullOr lib.types.str;
-        default = null;
-        description = ''
-          The path to the workdir.
+            If the filesystem is `neededForBoot`, this will be prefixed with `/sysroot`,
+            unless `underlyingDirectoriesOnRealRoot` is set to `true`.
 
-          This MUST be set if you set `upperdir`.
-        '';
+            This MUST be set if you set `upperdir`.
+          '';
+        };
+
+        underlyingDirectoriesOnRealRoot = lib.mkOption {
```

Comment:
`prefixWithSysroot` and `underlyingDirectoriesOnRealRoot` sound like synonyms to me so I don't know why you've inverted the logic?

Reply:
Hmm. I rather understand `underlyingDirectoriesOnRealRoot` as "do not automatically make these to be in the next root" (i.e. `/sysroot`). But maybe this is my lack of English skills. If you, as a native speaker, see it that way, I'll change it back.

Reply:
Oh, my brain imagines the initrd to be a temporary root, and the stage 2 file system to be the "real" root, since that's the root we really want.

Maybe `useStage1BaseDirectories` would be the most descriptive name.